rfc-106-docker-for-local-development.md

```diff
@@ -0,0 +1,37 @@
+# RFC 106: Use Docker for local development
+
+## Summary
+
+Adopt a Docker-based approach for local development instead of the Vagrant VM.
+
+## Problem
+
+We currently rely on the Vagrant development VM for the day to day dev cycle.
+
+- It has [many documented issues](https://docs.publishing.service.gov.uk/manual.html#development-vm)
+- Changes to puppet often result in a broken VM, because the infrastructure team don't use it.
+- We are adopting more cloud-based services into our stack, like S3 and Amazon's Elasticsearch. This means that our development environment doesn't look like production anymore, but we still need code in govuk-puppet.
+- It's hard to keep updated. Running the puppet command will often error and take a long time.
+- GOV.UK is moving to a containerised infrastructure
+
+## Possible approaches to local development
+
+- An approach based on **Docker and docker-compose**. Ben Thorner has a working prototype called [govuk-docker](https://github.com/benthorner/govuk-docker) and we have the [end-to-end tests](https://github.com/alphagov/publishing-e2e-tests) running in the same. Other GDS programmes like GOV.UK Pay use this approach.
+- An approach based on **GNU Guix** package manager, which Chris Baines uses via govuk-guix.
+- Because they are read-only, frontend applications on GOV.UK can be **run against production** using the startup.sh --live flag.
+- Developers on the GOV.UK PaaS work by developing against a **remote set of services** - we’d run many versions of GOV.UK in the cloud, which developers can use in development.
+- Applications can be run locally by **installing dependencies manually**. This is what some of GOV.UK’s Linux users do.
+
+
+Approach | Pro | Cons
+-- | -- | --
+Vagrant VM | Can look like production | Resource intensive, brittle
+Docker | Aligns with future hosting. Highly reproducible. | Slow because on a Mac it needs a VM. Development won’t match production exactly until we migrate Production onto a container-based platform.
```

Comment:
This has been a pain for me to. We're pretty greedy with our dependencies, but often we can afford to be more conservative. I've tried to capture this in a bunch of [needs and decisions](https://github.com/benthorner/govuk-docker/blob/master/NEEDS.md#running-services) when I started rolling my own.

Reply:
I'd like to suggest `gsp-local ` as an additional row for your table of approaches for completeness

[gsp-local](https://github.com/alphagov/gsp/blob/master/docs/gds-supported-platform/getting-started-gsp-local.md)  is also docker based and includes a local kubernetes node, Concourse CICD tooling, Docker Registy and Notary, Istio for networking, Prometheus for observability.  

It aligns with the [RE Strategy](https://reliability-engineering.cloudapps.digital/documentation/strategy-and-principles/re-strategy.html#2-build-a-new-common-hosting-platform-for-gds-services) however its still quite new.  We have several teams using it already and has recently undergone a round of user research.

Reply:
Thanks @pauldougan! I've added gsp-local to the list, and added an intention to move towards that in the future. I think it'd be a good approach to regard docker/docker-compose as the stepping stone towards GSP, since going from the current setup to Kubernetes would introduce many new technologies to GOV.UK at the same time.

Reply:
@thomasleese @benthorner I've added a note about dependencies to the RFC.